src/hotspot/cpu/riscv/stubGenerator_riscv.cpp

```diff
@@ -3661,6 +3661,915 @@ class StubGenerator: public StubCodeGenerator {
+    //    //   v16 = {f[t+4],e[t+4],b[t+4],a[t+4]}   " = v16[t+4] "
+    //    vsha2ch.vv v16, v17, v14
+    //
+    //    // Combine 2QW into 1QW
+    //    //
+    //    // To generate the next 4 words, "new_v10"/"v14" from v10-v13, vsha2ms needs
+    //    //     v10[0..3], v11[0], v12[1..3], v13[0, 2..3]
+    //    // and it can only take 3 vectors as inputs. Hence we need to combine
+    //    // v11[0] and v12[1..3] in a single vector.
+    //    //
+    //    // vmerge Vt4, Vt1, Vt2, V0
+    //    // Input
+    //    //  V0 = mask // first word from v12, 1..3 words from v11
+    //    //  V12 = {Wt-8, Wt-7, Wt-6, Wt-5}
+    //    //  V11 = {Wt-12, Wt-11, Wt-10, Wt-9}
+    //    // Output
+    //    //  Vt4 = {Wt-12, Wt-7, Wt-6, Wt-5}
+    //    vmerge.vvm v14, v12, v11, v0
+    //
+    //    // Generate next Four Message Schedule Words (hence allowing for 4 more rounds)
+    //    // Input
+    //    //  V10 = {W[t+ 3], W[t+ 2], W[t+ 1], W[t+ 0]}     W[ 3: 0]
+    //    //  V13 = {W[t+15], W[t+14], W[t+13], W[t+12]}     W[15:12]
+    //    //  V14 = {W[t+11], W[t+10], W[t+ 9], W[t+ 4]}     W[11: 9,4]
+    //    // Output (next four message schedule words)
+    //    //  v10 = {W[t+19],  W[t+18],  W[t+17],  W[t+16]}  W[19:16]
+    //    vsha2ms.vv v10, v14, v13
+    //
+    // BEFORE
+    //  v10 - v13 hold the message schedule words (initially the block words)
+    //    v10 = W[ 3: 0]   "oldest"
+    //    v11 = W[ 7: 4]
+    //    v12 = W[11: 8]
+    //    v13 = W[15:12]   "newest"
+    //
+    //  vt6 - vt7 hold the working state variables
+    //    v16 = {a[t],b[t],e[t],f[t]}   // initially {H5,H4,H1,H0}
+    //    v17 = {c[t],d[t],g[t],h[t]}   // initially {H7,H6,H3,H2}
+    //
+    // AFTER
+    //  v10 - v13 hold the message schedule words (initially the block words)
+    //    v11 = W[ 7: 4]   "oldest"
+    //    v12 = W[11: 8]
+    //    v13 = W[15:12]
+    //    v10 = W[19:16]   "newest"
+    //
+    //  v16 and v17 hold the working state variables
+    //    v16 = {a[t+4],b[t+4],e[t+4],f[t+4]}
+    //    v17 = {c[t+4],d[t+4],g[t+4],h[t+4]}
+    //
+    //  The group of vectors v10,v11,v12,v13 is "rotated" by one in each quad-round,
+    //  hence the uses of those vectors rotate in each round, and we get back to the
+    //  initial configuration every 4 quad-rounds. We could avoid those changes at
+    //  the cost of moving those vectors at the end of each quad-rounds.
+
+    //--------------------------------------------------------------------------------
+    // Quad-round 0 (+0, v10->v11->v12->v13)
+    __ vl1re32_v(v15, consts);
+    __ addi(consts, consts, 16);
+    __ vadd_vv(v14, v15, v10);
+    __ vsha2cl_vv(v17, v16, v14);
+    __ vsha2ch_vv(v16, v17, v14);
+    __ vmerge_vvm(v14, v12, v11);
+    __ vsha2ms_vv(v10, v14, v13); // Generate W[19:16]
+    //--------------------------------------------------------------------------------
+    // Quad-round 1 (+1, v11->v12->v13->v10)
+    __ vl1re32_v(v15, consts);
```

Comment:
I am still worried about the load latency if we do a `vl1re3_v` to get the consts for each round even for single pass.
Preloading the constants into vectors is less likely to have this issue, right? We should have enough vector registers for that purpose.

Reply:
Depending on hardware pipeline depth this load can actually be executed after
"__ vadd_vv(v14, v15, v10);" thus that instruction maybe already be retired when reaching round 1.

Preloading these, depending on the number of V-load ports, the preloading it self can be very cost as they can't be executed out-of-order in parallel.

So hiding the load in previous round can be faster, therefore my fast conclusion without numbers was at least for single pass no preloading *should* be better on bigger hardware.

I guess I need to get those numbers :)

Reply:
> Depending on hardware pipeline depth this load can actually be executed after "__ vadd_vv(v14, v15, v10);" thus that instruction maybe already be retired when reaching round 1.
> 
> Preloading these, depending on the number of V-load ports, the preloading it self can be very costly as they can't be executed out-of-order in parallel.

Make sense.

> So hiding the load in previous round can be faster, therefore my fast conclusion without numbers was at least for single pass no preloading _should_ be better on bigger hardware.

But I see that there is a true data dependence on the vector load for each round. Any thing I missed?
Say, for round2:
```
    // Quad-round 2 (+2, v12->v13->v10->v11)
    __ vl1re32_v(v15, consts);     ----> Define v15
    __ addi(consts, consts, 16);
    __ vadd_vv(v14, v15, v12);    ----> Use v15
```

Reply:
```
    __ vadd_vv(v14, v15, v11);
    <load can start>             <<----------------------------------------------------------------|
    __ vsha2cl_vv(v17, v16, v14);                                                                            |
    __ vsha2ch_vv(v16, v17, v14);                                                                          |
    __ vmerge_vvm(v14, v13, v12);                                                                        |
    __ vsha2ms_vv(v11, v14, v10); // Generate W[23:20]                                   |
    //--------------------------------------------------------------------------------                        |
    // Quad-round 2 (+2, v12->v13->v10->v11)                                                   |
    __ vl1re32_v(v15, consts); -----------------------------------------------------------------
```

No ?

Reply:
If you consider register renaming load should enable to start even earlier.
E.g. load into vX, then rename vX to v15 after the add that uses v15.

Reply:
We don't either have such hardware, we simulate via gem5.
Ventana v2 should have 15 wide pipeline with RVV 1.0 how knows how this will execute on such :)

As we don't know I think you are correct in we should write the most readable version first.
And later we can apt these for hwprobe triplet of vendor/arch/impl if we think that it's worth it.

Reply:
Not yet addressed.

Reply:
As I don't have access to these instruction in performance simulator.
I'll leave this as is, since now 256/512 are identical in that regard.